trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/TrinoLogsResource.java

```diff
@@ -0,0 +1,242 @@
+                logStreamNamesPasses(entry, getLogEvents);
+        int limit = getLogEvents.limit.orElse(Integer.MAX_VALUE);
+
+        List<SaveEntry> filteredEntries = loggerController.savedEntries(startFromHead, predicate);
+
+        record StartIndex(TokenType tokenType, int index) {}
+
+        // next token is merely an index
+        StartIndex startIndex = getLogEvents.nextToken().map(nextToken -> {
+            try {
+                TokenType tokenType;
+                if (nextToken.startsWith(TokenType.FORWARDS.prefix)) {
+                    tokenType = TokenType.FORWARDS;
+                }
+                else if (nextToken.startsWith(TokenType.BACKWARDS.prefix)) {
+                    tokenType = TokenType.BACKWARDS;
+                }
+                else {
+                    throw new WebApplicationException("Invalid nextToken", BAD_REQUEST);
+                }
+                return new StartIndex(tokenType, Integer.parseInt(nextToken.substring(2)));
+            }
+            catch (NumberFormatException _) {
+                throw new WebApplicationException("Invalid nextToken", BAD_REQUEST);
+            }
+        }).orElseGet(() -> new StartIndex(TokenType.FORWARDS, 0));
+
+        List<Event> events = filteredEntries.stream()
+                .skip(startIndex.index)
+                .limit(limit)
+                .map(entry -> new Event("trino", entry.entryId(), entry.timestamp().toEpochMilli(), entry.timestamp().toEpochMilli(), format(entry)))
+                .collect(toImmutableList());
+
+        Optional<String> nextBackwardToken;
+        Optional<String> nextForwardToken;
+        if (startIndex.tokenType == TokenType.FORWARDS) {
+            nextBackwardToken = TokenType.BACKWARDS.atIndex(startIndex.index);
+            nextForwardToken = ((startIndex.index + limit) < filteredEntries.size()) ? TokenType.FORWARDS.atIndex(startIndex.index + limit) : TokenType.FORWARDS.atIndex(startIndex.index);
+        }
+        else {
+            nextBackwardToken = ((startIndex.index - limit) >= 0) ? TokenType.BACKWARDS.atIndex(startIndex.index - limit) : TokenType.BACKWARDS.atIndex(startIndex.index);
+            nextForwardToken = TokenType.FORWARDS.atIndex(startIndex.index);
+        }
+
+        GetLogEventsResponse response = new GetLogEventsResponse(events, nextBackwardToken, nextForwardToken);
+
+        return Response.ok(response).build();
+    }
+
+    private static boolean logStreamNamesPasses(SaveEntry entry, GetLogEventsRequest getLogEvents)
+    {
+        return getLogEvents.logStreamName.map(logStreamName -> logStreamNamePasses(entry, logStreamName))
+                .orElseGet(() -> getLogEvents.logStreamNames
+                        .map(logStreamNames -> logStreamNames.stream().anyMatch(logStreamName -> logStreamNamePasses(entry, logStreamName)))
+                        .orElse(true));
+    }
+
+    private static boolean logStreamNamePasses(SaveEntry entry, String logStreamName)
```

Comment:
Nit: could we swap the order of the arguments to this function? It's the other way around from the other 2 filter